### PR TITLE
Estimated SmartFee from BitcoinCore

### LINKED ISSUE #629: Receive 3rd Party Feed of Transaction Fees
* [ ] Receive 3rd party feed of current mainnet (and ideally testnet too) transaction fees so we can track the real cost of our upcoming transactions.
* [ ] Add to Health Endpoint

### PR DESCRIPTION
**Pull request by jwicks31**
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

Resolves #629  

- Added `refreshSmartFee` to `Health/HealthController` - Gets the approximate fee per kilobyte needed for a transaction to begin confirmation within a number of blocks(Configurable with ENV variable) and updates MongoDb.
- Added `feeEstimateBitcoinConfMinTargetBlock` to `Configuration`, which is the number of blocks supplied to Bitcoin-Core's estimateSmartFee.
- Updated Health Functional Tests

### CODE REVIEW

File: src/Configuration.ts [hunk 1]
```diff
@@ -103,6 +104,7 @@ const defaultConfiguration: Configuration = {
 
   healthIntervalInSeconds: 30,
   lowWalletBalanceInBitcoin: 1,
+  feeEstimateBitcoinConfMinTargetBlock: 1,
```
Comment (lautarodragan) [inline src/Configuration.ts]: Maybe just `feeEstimateMinTargetBlock`? (without the `BitcoinConf`)

File: src/Health/HealthController.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@ enum LogTypes {
 
 export interface HealthControllerConfiguration {
   readonly lowWalletBalanceInBitcoin: number
+  readonly feeEstimateBitcoinConfMinTargetBlock: number
```
Comment (lautarodragan) [inline src/Health/HealthController.ts]: Same here?

File: src/Health/HealthController.ts [hunk 3]
```diff
@@ -146,6 +152,23 @@ export class HealthController {
     return networkInfo
   }
 
+  private async getEstimatedSmartFee(): Promise<EstimatedSmartFeeInfo> {
+    return await this.bitcoinCore.estimateSmartFee(this.configuration.feeEstimateBitcoinConfMinTargetBlock)
+  }
+
+  private async updateEstimatedSmartFee(estimatedSmartFeeInfo: any): Promise<EstimatedSmartFeeInfo> {
```
Comment (lautarodragan) [inline src/Health/HealthController.ts]: Can this `any` be avoided?

Comment (jwicks31) [reply src/Health/HealthController.ts]: Yep, missed changing that one.

File: src/Configuration.ts (hunk 1 above)
Comment (jwicks31) [reply src/Configuration.ts]: Sounds good!

**Comment by jwicks31** [comment]
@lautarodragan I made the changes. I also merged in master which had the HealthDAO, so I made use of that for the calls to mongo here as well.

File: tests/functional/endpoints.ts [hunk 4]
```diff
@@ -42,7 +42,7 @@ describe('Health Endpoint Returns the Correct Fields', async (assert: any) => {
     const response = await getHealth(NODE_PORT)
     const healthData = await response.json()
 
-    const { mongoIsConnected, ipfsInfo, blockchainInfo, networkInfo, walletInfo } = healthData
+    const { mongoIsConnected, ipfsInfo, blockchainInfo, networkInfo, walletInfo, estimatedSmartFeeInfo } = healthData
```
Comment (lautarodragan) [inline tests/functional/endpoints.ts]: You added the `estimatedSmartFeeInfo` but nowhere else I think... is this missing an assertion on it?

**Comment by lautarodragan** [comment]
@jwicks31 tests failing in Travis due to

```
[2018-11-01T01:12:00.670Z] ERROR (164 on 7376c9ba5c5e): Failed to getHealthInfo
    module: "Health"
    file: "Router"
    method: "onGetHealth"
    error: {
      "cause": {
        "errno": "ECONNREFUSED"
        "code": "ECONNREFUSED"
        "syscall": "connect"
        "address": "172.18.0.6"
        "port": 18443
        "stack": [
          "Error: connect ECONNREFUSED 172.18.0.6:18443"
          "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"
        ]
        "message": "connect ECONNREFUSED 172.18.0.6:18443"
        "type": "Error"
      }
      "isOperational": true
      "errno": "ECONNREFUSED"
      "code": "ECONNREFUSED"
      "syscall": "connect"
      "address": "172.18.0.6"
      "port": 18443
      "stack": [
        "Error: connect ECONNREFUSED 172.18.0.6:18443"
        "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"
      ]
      "message": "connect ECONNREFUSED 172.18.0.6:18443"
      "type": "OperationalError"
    }
[2018-11-01T01:12:01.672Z] ERROR (164 on 7376c9ba5c5e): Failed to getHealthInfo
    module: "Health"
    file: "Router"
    method: "onGetHealth"
    error: {
      "message": "Loading wallet..."
      "code": -28
      "name": "RpcError"
      "stack": [
        "RpcError: -28 Loading wallet..."
        "    at getRpcResult (/usr/src/app/node_modules/bitcoin-core/dist/src/parser.js:39:11)"
        "    at Client.rpc (/usr/src/app/node_modules/bitcoin-core/dist/src/parser.js:82:14)"
        "    at Client.tryCatcher (/usr/src/app/node_modules/bluebird/js/release/util.js:16:23)"
        "    at Promise._settlePromiseFromHandler (/usr/src/app/node_modules/bluebird/js/release/promise.js:512:31)"
        "    at Promise._settlePromise (/usr/src/app/node_modules/bluebird/js/release/promise.js:569:18)"
        "    at Promise._settlePromise0 (/usr/src/app/node_modules/bluebird/js/release/promise.js:614:10)"
        "    at Promise._settlePromises (/usr/src/app/node_modules/bluebird/js/release/promise.js:694:18)"
        "    at Promise._fulfill (/usr/src/app/node_modules/bluebird/js/release/promise.js:638:18)"
        "    at Request._callback (/usr/src/app/node_modules/bluebird/js/release/nodeback.js:45:21)"
        "    at Request.self.callback (/usr/src/app/node_modules/request/request.js:185:22)"
        "    at Request.emit (events.js:187:15)"
        "    at Request.EventEmitter.emit (domain.js:442:20)"
        "    at Request.<anonymous> (/usr/src/app/node_modules/request/request.js:1161:10)"
        "    at Request.emit (events.js:182:13)"
        "    at Request.EventEmitter.emit (domain.js:442:20)"
        "    at IncomingMessage.<anonymous> (/usr/src/app/node_modules/request/request.js:1083:12)"
      ]
      "type": "RpcError"
    }
```

I'm guessing this is just the bitcoin node / wallet taking longer than expected until responsive. @warrenv can you lend a hand here?

File: tests/functional/endpoints.ts (hunk 4 above)
Comment (jwicks31) [reply tests/functional/endpoints.ts]: Yeah I must have accidentally deleted it whenever I merged in master. I'll add it back

**Comment by warrenv** [comment]
yes, it looks like the bitcoind is not ready yet. It might need a slightly larger delay than what was initially used on a local development machine.

File: tests/functional/endpoints.ts (hunk 4 above)
Comment (lautarodragan) [reply tests/functional/endpoints.ts]: Ouch :/ Merges can be a PITA

**Comment by jwicks31** [comment]
@warrenv I increased the delay to 10s(from 5s) and it's still throwing the same error. Do you think that should have been enough to resolve it?

**Comment by warrenv** [comment]
that should have been enough. you'll have to get inside the poet-node container and check the env vars are set correctly.

Does it work when you run it locally? (Sorry, not at the computer right now).

**Comment by jwicks31** [comment]
@warrenv it does the same thing locally:
```
ERROR (148 on d4dbf1dee688): Failed to getHealthInfo
    module: "Health"
    file: "Router"
    method: "onGetHealth"
    error: {
      "cause": {
        "errno": "ECONNREFUSED",
        "code": "ECONNREFUSED",
        "syscall": "connect",
        "address": "172.22.0.4",
        "port": 18443,
        "stack": [
          "Error: connect ECONNREFUSED 172.22.0.4:18443",
          "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"
        ],
        "message": "connect ECONNREFUSED 172.22.0.4:18443",
        "type": "Error"
      },
      "isOperational": true,
      "errno": "ECONNREFUSED",
      "code": "ECONNREFUSED",
      "syscall": "connect",
      "address": "172.22.0.4",
      "port": 18443,
      "stack": [
        "Error: connect ECONNREFUSED 172.22.0.4:18443",
        "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"
      ],
      "message": "connect ECONNREFUSED 172.22.0.4:18443",
      "type": "OperationalError"
    }
```

The tests still pass and everything(same as through travis), but shows this error.

**Comment by warrenv** [comment]
Fixed here: https://github.com/poetapp/node/pull/709

**Pull request merged**